### PR TITLE
Pa02

### PR DESCRIPTION
**Pull request by nimve**
Here is our submission for Pa02. Please take a look, and let us know if changes should be made. Best regards, Trude Haug Almestrand and Nina Mariann Vesseltun

### CODE REVIEW

**Review by trudea — APPROVED** [review]

File: src/pa02/snakes_simulation.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+        :return: Updates class variable position
+        """
+        self.position += randint(1, 6) + self.added_steps
+        remembered_position = self.position
+        self.position += self.board.position_adjustment(self.position)
+        if self.position < remembered_position:
+            self.added_steps = self.extra_steps
+        else:
+            self.added_steps = 0
```
Comment (huynhngoc) [inline src/pa02/snakes_simulation.py:86]: 👍  good!

File: src/pa02/snakes_simulation.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        """
+        dice = randint(1, 6)
+        if dice - self.steps_back >= 0:
+            self.position += dice - self.steps_back
+        remembered_position = self.position
+        self.position += self.board.position_adjustment(self.position)
+        if self.position > remembered_position:
+            self.steps_back = self.dropped_steps
+        else:
+            self.steps_back = 0
```
Comment (huynhngoc) [inline src/pa02/snakes_simulation.py:108]: Nice one here also 🌷

File: src/pa02/snakes_simulation.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+class Simulation:
+    def __init__(self, player_field, board=None, seed=None,
+                 randomize_players=True):
+        self.player_field = player_field
+        self.board = board
+        self.results = []
+        if not self.board:
+            self.board = Board()
+        self.seed = seed
```
Comment (huynhngoc) [inline src/pa02/snakes_simulation.py:119]: `random.seed(seed)` here.

File: src/pa02/snakes_simulation.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            self.steps_back = self.dropped_steps
+        else:
+            self.steps_back = 0
+
+
+class Simulation:
+    def __init__(self, player_field, board=None, seed=None,
+                 randomize_players=True):
+        self.player_field = player_field
+        self.board = board
```
Comment (huynhngoc) [inline src/pa02/snakes_simulation.py:115]: Use default board here if board is None

File: src/pa02/snakes_simulation.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+            self.board = Board()
+        self.seed = seed
+        self.randomize_players = randomize_players
+
+    def single_game(self):
+        """
+
+        :return: A tuple of winning number of moves and player type of winner
+        """
+        moves = 0
```
Comment (huynhngoc) [inline src/pa02/snakes_simulation.py:127]: This should be controlled inside Player, within the `move` method.

File: src/pa02/test_pa02.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                         (65, 82), (68, 85)]
+    assert b.goal == 90
+
+
+def test_goal_reached():
+    """Checking if goal_reached works"""
+    b = cs.Board()
+    assert b.goal_reached(90) is True
+    assert b.goal_reached(13) is False
+    assert b.goal_reached(100) is True
```
Comment (huynhngoc) [inline src/pa02/test_pa02.py:25]: Please have the habit of omitting `is True` and using `not` instead of `is False`

File: src/pa02/test_pa02.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+def test_lazy_player():
+    """Checking if LazyPlayer behaves as desired"""
+    random.seed(999)
+    b = cs.Board()
+    p = cs.LazyPlayer(b)
+    p.position = 2
+    p.move()
+    random.seed(999)
+    p.move()
+    assert p.position == 15
```
Comment (huynhngoc) [inline src/pa02/test_pa02.py:68]: Not really how we control randomness in real-life, but nice try.

File: src/pa02/test_pa02.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    later_results = s.results
+    assert first_results != later_results
+
+
+def test_get_results():
+    """Checking if get_results works"""
+    s = cs.Simulation(player_field=[cs.Player, cs.ResilientPlayer,
+                                    cs.LazyPlayer], seed=999)
+    s.run_simulation(2)
+    assert s.get_results() == s.results
```
Comment (huynhngoc) [inline src/pa02/test_pa02.py:117]: Not really convincing here as this is what's in your code.

Comment (huynhngoc) [reply src/pa02/test_pa02.py:117]: Which makes this test case has no values.
Testing get_results can just test for:
- Return values: list of tuple
- 1st item in tuple: an int that is greater than 0
- 2nd item in the tuple: a str that is in 'Player', 'ResilientPlayer' or 'LazyPlayer'

Comment (huynhngoc) [reply src/pa02/test_pa02.py:117]: If possible, you can make a custom board that have ladder from 1 to 90, 2 to 90, 3 to 90, 4 to 90, 5 to 90, 6 to 90.
Then, any player that plays first wins. And you can explicitly specify the expected return value for get_results.

**Review by huynhngoc — APPROVED** [review]
Approved.

Please note that we expect you both commit regularly and switch roles regularly so that there are similar contributions (and commits) from both members to prove that you two both work on the project.
Remember to commit and switch role more often in the Jan/Jun block (or they will fail you).

**Pull request merged**